feat(argo-cd): Add argocd-cmd-params-cm

pdrastil opened the pull request:
This PR should start discussion regarding https://github.com/argoproj/argo-helm/issues/1210

Changes:
- Added argocd-cmd-params-cm configmap
- Exposed all known config options via environment variables
- Rewire internal parameters for Redis / Dex
- Added deprecation notice for arguments that were superseded by new config

Resolves:
- https://github.com/argoproj/argo-helm/issues/1210

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo).
* [x] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

Changes are automatically published when merged to `master`. They are not published on branches.

pdrastil commented on charts/argo-cd/README.md.gotmpl:
```diff
@@ -39,6 +39,9 @@ repoServer:
     minReplicas: 2
+
+applicationSet:
+  replicaCount: 2
```
Shouldn't this be also changed to `replicas` as everywhere else?

mkilchhofer replied:
I think both new components of 2.3 (notification & appset controller) does not yet support HA?

mkilchhofer commented on charts/argo-cd/templates/argocd-configs/argocd-cmd-params-cm.yaml:
```diff
@@ -0,0 +1,106 @@
+
+  ## Controller Properties
+  # Number of application status processors (default 20)
+  controller.status.processors: "{{ .Values.controller.args.statusProcessors }}"
```
I think we should prepare a breaking change (major version) and refactor it. It looks kinda strange to do a random variable mapping here. Eg. `controller.status.processors = controller.args.statusProcessors`


I'd prefer a structure like:
```yaml
configs:
  cm: # or config: / configmap: / ...?
    # <keys that goes into argocd-cm>

  params:
    # <keys that goes into argocd-cmd-params-cm>
```
Maybe we can use a template function for that and merge the resulting yaml with our templated things like 

- `dex.server: "http://{{ template "argo-cd.dex.fullname" . }}:{{ .Values.dex.servicePortHttp }}"`
- `  redis.server: "{{ template "argo-cd.redis.fullname" . }}:{{ .Values.redis.servicePort }}"`
- etc

pdrastil replied:
I'm open about such change. If I understand correctly you would like to consolidate everything under `configs` section including stuff that is currently under `server`? And after that merge user config with internally set keys such as dex, redis, etc.?

mkilchhofer reviewed (commented):
I really like the intent of implementing #1210 but I'd think about refactoring the chart completely as there are already deprecations in your proposal.

pdrastil commented on charts/argo-cd/values.yaml:
```diff
@@ -1694,6 +1702,23 @@ repoServer:
+  # Provide cache expirations
+  cache:
+    # -- Default cache expiration
+    default: 24h
```
What do you want to do with `cache` and `repoServer` config? I've seen that this is cross-cutting concern where same value repeats for multiple components and I'm not sure if we want to somehow simplify this or we want to keep it dry and somehow merge / template 1 section into params as well.

pdrastil commented:
@mkilchhofer Please take a look on another iteration and tell me WHYT. When you  wanted to go with breaking change I've tried to consolidate parameters such as `tlsCerts` and `tlsCertsAnnotations` under one config section. Things I haven't visited yet:

- credentials / clusters / templates - Do we want to have special config section for these? What to do with legacy configuration option that is deprecated? remove this one?
- notifications - it have configuration in `argocd-notifications-cm` in upstream repo - maybe it should be also refactored and added to configs
- `configs.cm.url` - could we use this for notification context and maybe ingress host as a default?

mkilchhofer commented:
Asked the community in the slack channel:
https://cloud-native.slack.com/archives/C02097VUKUM/p1654880501143779

pdrastil commented:
> Asked the community in the slack channel for their thoughts about this refactoring: https://cloud-native.slack.com/archives/C02097VUKUM/p1654880501143779
> 
> > <img alt="grafik" width="1147" src="https://user-images.githubusercontent.com/7290987/173118013-e6b68c70-84d1-40eb-b984-a5c8aa6e44c9.png">

Thanks @mkilchhofer for the info - refactoring is not fully done because I'm waiting for the feedback. I plan to finish remaining sections as well if we are fine with the way this PR is heading.

ArgTang reviewed (changes requested):
missing migration guide in readme. 
* it would be nice if one is able to override a components loglevel for debugging. Here you can only set the loglevel globally

pdrastil commented on charts/argo-cd/templates/argocd-repo-server/deployment.yaml:
```diff
@@ -44,24 +45,91 @@ spec:
+          valueFrom:
+            configMapKeyRef:
+              name: argocd-cmd-params-cm
+              key: reposerver.log.level
```
@ArgTang Log level / format is now controlled via config map. Just provide proper key in params section for given component.

ArgTang replied:
ah you are right. thanks for the correction :)

pdrastil commented on charts/argo-cd/templates/argocd-applicationset/deployment.yaml:
```diff
@@ -41,7 +41,7 @@ spec:
             - applicationset-controller
```
We can align this.

pdrastil replied:
@ArgTang Added to refactoring.

pdrastil commented on charts/argo-cd/values.yaml:
```diff
@@ -17,6 +17,29 @@ global:
+  imagePullSecrets: []
+
+  logging:
+    # -- Set the global logging format. One of: `text` or `json`
```
Waiting for 2.4.x release - I've seen that upstream added extra flags so these components should be configurable via env, but now it's possible only via command line args.

ArgTang replied:
imo it would be weird if the global logsetting dont work on some components. I would prefer using them as cmd flags if you want to release this before upstream additions.

pdrastil commented on charts/argo-cd/values.yaml:
```diff
@@ -17,6 +17,29 @@ global:
+    # -- Set the global logging format. One of: `text` or `json`
+    format: text
+    # -- Set the global logging level. One of: `debug`, `info`, `warn` or `error`
+    level: info
```
@mkilchhofer WDYT? Keep current info level or change to warn?

mkilchhofer replied:
Hmmmm I'd keep info and let the user choice :)

pdrastil commented on charts/argo-cd/templates/argocd-server/clusterrole.yaml:
```diff
@@ -27,7 +27,7 @@ rules:
     verbs:
       - get
-  {{- if eq (index .Values.server.config "exec.enabled") "true" }}
+  {{- if eq (index .Values.configs.cm "exec.enabled") "true" }}
```
@mkilchhofer Why was this added into Cluster admin role for Argo CD? Do we expect that admin shouldn't be able to check logs?

mkilchhofer replied:
It will be turned on by in future releases. Argoproj core team decided to disable this by default to not break users setups.
And also in my current position at Swiss Post, our security engineering decided to not enabled that capability on prod clusters.

pdrastil replied:
I see - it makes sense from this perspective.

mkilchhofer commented on charts/argo-cd/README.md.gotmpl:
```diff
@@ -28,7 +28,7 @@ redis-ha:
 
 controller:
-  enableStatefulSet: true
+  replicas: 2
```
2 replica for the application controller is a bad default. The controller cannot shard its work if Argo CD manages only one cluster.

The shard number of a managed cluster is calculated via hostname:
https://github.com/argoproj/argo-cd/blob/v2.4.0/cmd/util/cluster.go#L133

pdrastil replied:
I guess number of replicas comes from this example: https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller. Do you want to use 3 for HA setups?

mkilchhofer commented on charts/argo-cd/README.md.gotmpl:
```diff
@@ -48,16 +51,16 @@ redis-ha:
 
 controller:
-  enableStatefulSet: true
+  replicas: 2
```
use 1 as default. See other comment

pdrastil replied:
1 is not actually great for HA example

mkilchhofer replied:
The controller simply does not support HA.
Ref: https://github.com/argoproj/argo-cd/pull/8340/files?diff=unified&w=0&short_path=5ab05fa#diff-5ab05fafb9f84e2256b1b32e91f4de69e89e5eba19eae3818c7aba539fb22fb0

> Currently, the application-controller does not provide high availability to support workload distribution. This causes conflicts and leads to orphaned/unmanaged applications in scenarios where the application controller handling these resources malfunctions/crashes.

pdrastil replied:
Good to know - I always thought it can handle failure. Anyway updated to 1

mkilchhofer commented on charts/argo-cd/README.md.gotmpl:
```diff
@@ -91,10 +94,38 @@ Please use `kubectl` to upgrade CRDs manually from [crds](crds/) folder or via t
+- Renamed parameters to align with rest of the chart
+  - `replicaCount` to `replicas`
+  - `securityContext` to `containerSecurityContext`
+  - `extraEnv` to `env`
+  - `extraEnvFrom` to `envFrom`
```
Maybe it's time to analyze the other charts and consolidate the naming. I merged one PR for rollouts with extraEnv couple of hours ago :(

pdrastil replied:
Agree one same naming convention. All original components (server, reposerver and controller) were using env and envFrom so I tried to stick with that. If you want to change these to extraEnv it is also fine. Imo I believe env will be used less often because of params-cm.

mkilchhofer replied:
It's almost 50:50 how the current usage is:
```console
mkilchhofer@Kicm-MBP:~/git/github/argo-helm$ git grep Env: | grep values.yaml
charts/argo-cd/values.yaml:  extraEnv: []
charts/argo-cd/values.yaml:  extraEnv: []
charts/argo-rollouts/values.yaml:  extraEnv: []
charts/argo-rollouts/values.yaml:  extraEnv: []
charts/argo-workflows/values.yaml:  extraEnv: []
charts/argo-workflows/values.yaml:  extraEnv: []
charts/argocd-image-updater/values.yaml:extraEnv: []

mkilchhofer@Kicm-MBP:~/git/github/argo-helm$ git grep env: | grep values.yaml
charts/argo-cd/ci/ha-static-values.yaml:  env:
charts/argo-cd/values.yaml:  env:
charts/argo-cd/values.yaml:  env: []
charts/argo-cd/values.yaml:  env: []
charts/argo-cd/values.yaml:  env: []
charts/argo-cd/values.yaml:  #   env:
charts/argo-cd/values.yaml:  env: []
charts/argo-events/values.yaml:  env: []
charts/argo-events/values.yaml:  env: []
charts/argo-workflows/values.yaml:  env: {}
charts/argo-workflows/values.yaml:  env: {}
charts/argocd-image-updater/values.yaml:    #   credentials: env:REGISTRY_SECRET
```

But its really inconsistent 😿 even in one chart it is different 🤣

mkilchhofer replied:
Lets use `env:` and `envFrom:`

mkilchhofer commented on charts/argo-cd/values.yaml:
```diff
@@ -1752,6 +1654,258 @@ repoServer:
 configs:
+  # General Argo CD configuration
+  ## Ref: https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml
+  # TODO: Should we add more examples from link above / keep example for oidc.config?
```
Yes, why not removing unrelevant settings. The less examples, the less work needed to have a look into when there's a new release coming in.

mkilchhofer commented on charts/argo-cd/values.yaml:
```diff
@@ -1752,6 +1654,258 @@ repoServer:
+    ## Custom secrets. Useful for injecting SSO secrets into environment variables.
+    ## Ref: https://argoproj.github.io/argo-cd/operator-manual/sso/
+    ## Note that all values must be non-empty.
+    extra:
+      {}
```
```suggestion
    extra: {}
```

What about unify how empty dicts are defined across this file?

pdrastil replied:
Done

zestrells commented:
Hi everyone, is someone taking a look at adding this config `argocd-cmd-params-cm.yaml` to repo-server? We are getting ready to leverage argocd in production and want to ensure we can change the `reposerver.parallelism.limit` value.

pdrastil commented:
> Hi everyone, is someone taking a look at adding this config `argocd-cmd-params-cm.yaml` to repo-server? We are getting ready to leverage argocd in production and want to ensure we can change the `reposerver.parallelism.limit` value.

@zestrells I'm currently pulling backward comapatible changes from this PR and trying to push it in smaller pieces.

jmeridth commented:
@pdrastil #1320 has been merged.  `controller.args.appHardResyncPeriod` can be added as deprecated to this PR.  Thank you.

jmeridth commented on charts/argo-cd/templates/NOTES.txt:
```diff
@@ -1,3 +1,40 @@
+{{- if .Values.controller.args.appResyncPeriod }}
+DEPRECATED option controller.args.appResyncPeriod - Use configs.params.timeout.reconciliation
+{{- end }}
+{{- if .Values.controller.args.appHardResyncPeriod }}
```
❤️

jmeridth commented on charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:
```diff
@@ -38,37 +38,44 @@ spec:
-        - {{ .Values.controller.args.appResyncPeriod | quote }}
+        - {{ . | quote }}
+        {{- end }}
+        {{- with .Values.controller.args.appHardResyncPeriod }}
```
❤️

jmeridth reviewed (commented):
Thank you for adding this

mkilchhofer commented on charts/argo-cd/templates/argocd-server/deployment.yaml:
```diff
@@ -222,10 +378,10 @@ spec:
+      {{- with .Values.server.initContainers }}
       initContainers:
-      {{- toYaml .Values.server.initContainers | nindent 6 }}
+      {{- toYaml . | nindent 6 }}
```
```suggestion
        {{- toYaml . | nindent 6 }}
```

pdrastil replied:
Done

mkilchhofer commented on charts/argo-cd/templates/argocd-server/deployment.yaml:
```diff
@@ -33,55 +33,209 @@ spec:
-      {{- if .Values.global.securityContext }}
-      securityContext: {{- toYaml .Values.global.securityContext | nindent 8 }}
+      {{- with .Values.global.securityContext }}
+      securityContext: {{- toYaml . | nindent 8 }}
```
Not a blocker but the container-level security context is defined like this
```suggestion
      securityContext:
        {{- toYaml . | nindent 8 }}
```

pdrastil replied:
Done

mkilchhofer commented on charts/argo-cd/templates/argocd-repo-server/deployment.yaml:
```diff
@@ -44,40 +44,134 @@ spec:
-          {{- include "argo-cd.redisPasswordEnv" . | nindent 8 }}
           {{- with .Values.repoServer.env }}
-          {{- toYaml . | nindent 8 }}
+            {{- toYaml . | nindent 8 }}
```
Please reduce the indent of the env block by 2. We should use the same style inside one yaml file. All lists are defined like this:
```yaml
key:
- listitem1
- listitem2
```
and not like this:
```yaml
key:
  - listitem1
  - listitem2
```

pdrastil replied:
Done

mkilchhofer commented on charts/argo-cd/templates/argocd-server/deployment.yaml:
```diff
@@ -33,55 +33,209 @@ spec:
+          {{- toYaml . | nindent 10 }}
         {{- end }}
-        {{- if or .Values.server.env .Values.externalRedis.password .Values.externalRedis.existingSecret }}
         env:
```
Reduce indent block by 2

pdrastil replied:
Done

mkilchhofer commented on charts/argo-cd/values.yaml:
```diff
@@ -123,26 +123,29 @@ controller:
   replicas: 1
 
   ## Application controller commandline flags
```
```suggestion
  # -- DEPRECATED Application controller commandline flags
```

pdrastil replied:
I would like to keep this for refactoring and drop all `extraEnv` and `extraArgs` and keep just `args` and `env`.

mkilchhofer replied:
Then we need to add something different. Currently the key is undocumented in README (line 404):
```
| controller.args | object | `{}` | |
```

pdrastil replied:
OK marked as deprecated and then we will change the meaning during the recatoring.

pdrastil commented:
@mkilchhofer any update on this?

jmeridth reviewed (commented):
LGTM 👍

nice cleanup of `if` to `with`
